cpp/src/arrow/acero/swiss_join.cc

```diff
@@ -2520,11 +2518,6 @@ class SwissJoin : public HashJoinImpl {
       return status();
     }
 
-    if (!local_states_[thread_index].hash_table_ready) {
-      local_states_[thread_index].hash_table_ready = hash_table_ready_.load();
-    }
-    ARROW_DCHECK(local_states_[thread_index].hash_table_ready);
```

Comment:
Upper level logic https://github.com/apache/arrow/blob/1e45e182a71d51223cfb82b236dde07b436f44f1/cpp/src/arrow/acero/hash_join_node.cc#L1039 guarantees the readiness of the hash table by calling the interfaces (`impl_->BuildHashTable` and `impl_->ProbeSingleBatch`) in sequence so we don't actually have to do it all over.